Add x-pack plugin for new platform server licensing information

eliperelman opened the pull request:
## Summary

Add x-pack plugin for new platform server licensing information. This will eventually replace the licensing information consumed via `xpack_main`.

- [x] Server implementation
- [x] Server tests

Fixes #18356.
Supersedes #41026.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [x] ~~This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] ~~[Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] ~~This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [x] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

### Dev Docs

Add x-pack plugin for new platform server licensing information. This will eventually replace the licensing information consumed via `xpack_main`. Upon setup, this plugin exposes an observable API for inspecting and making checks against the license information.

```ts
license$.subscribe(license => {
   console.log(license.uid);
   console.log(license.isActive);
   console.log(license.type);

   const { check } = license.check('my-plugin', LICENSE_TYPE.gold);

   if (check !== LICENSE_STATUS.Valid) {
     disableSomething();
   }
});
```

elasticmachine commented:
Pinging @elastic/kibana-platform

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13447/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13615/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13628/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13634/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13687/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14032/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14070/)

joshdover commented on x-pack/plugins/licensing/server/index.ts:
```diff
@@ -0,0 +1,14 @@
+
+import { PluginInitializerContext } from 'src/core/server';
+import { schema } from './schema';
+import { LicensingPlugin } from './licensing_plugin';
```
nit: [by convention](https://github.com/elastic/kibana/blob/master/src/core/CONVENTIONS.md), this file should be called "plugin"

joshdover commented on x-pack/plugins/licensing/server/types.ts:
```diff
@@ -0,0 +1,43 @@
+  message?: string;
+}
+/** @public */
+export interface ILicensingPluginSetup {
```
I find the name of this a little off from the pattern we've established. The convention we've been using in core is that the name of the interface returned from `setup` is called `*Setup`. That's close to what this service does, but not quite.

I also think we should make the return value of `setup` be more flexible so that we can add additional properties without it being a breaking change.

This is how I envision it:

```ts
interface LicensePluginSetup {
  // `ILicense` would be the name of the interface is currently called `ILicensePluginSetup`
  license$: Observable<ILicense>;
}

class LicensePlugin implements Plugin<LicensePluginSetup> {
  setup() {
    return { license$ };
  }
}
```

Also, can we add TSDocs to each property of this interface?

joshdover commented on x-pack/plugins/licensing/server/license_feature.ts:
```diff
@@ -0,0 +1,38 @@
+import { LicensingPluginSetup } from './licensing_plugin_setup';
+import { LicenseFeatureSerializer } from './types';
+
+export class LicenseFeature {
```
Can we add a basic test for this?

joshdover commented on x-pack/plugins/licensing/server/licensing.test.ts:
```diff
@@ -0,0 +1,275 @@
+  clusterClient: jest.Mocked<PublicMethodsOf<ClusterClient>>;
+}
+
+describe('licensing', () => {
```
Can we split up these tests? I think each file should have it's own file of unit tests that are pretty isolated.

eliperelman replied:
These tests are all for the setup license interface, all a single file.

joshdover commented on x-pack/plugins/licensing/server/licensing_plugin_setup.ts:
```diff
@@ -0,0 +1,178 @@
+  private readonly hasLicense: boolean;
+  private readonly license: any;
+  private readonly features: any;
+  private _signature!: string;
```
any reason this is prefixed with an underscore and the other mutables values are not?

eliperelman replied:
There is already a getter for the signature, this is the backing field for it.

joshdover commented on x-pack/plugins/licensing/server/licensing_plugin_setup.ts:
```diff
@@ -0,0 +1,178 @@
+  private readonly features: any;
+  private _signature!: string;
+  private objectified!: any;
+  private featuresMap: Map<string, LicenseFeature>;
```
nit: You could mark this field as readonly since the Map instance itself should never change.

joshdover commented on x-pack/plugins/licensing/server/licensing_plugin.ts:
```diff
@@ -0,0 +1,128 @@
+import { LicensingConfig } from './licensing_config';
+import { LicensingPluginSetup } from './licensing_plugin_setup';
+
+export class LicensingPlugin implements Plugin<LicensingPluginSubject> {
```
There's a lot going on in this class. I think it might be better to move the polling code into a sub-service that this class simply wires up and exposes. This would make adding features to this class and testing this class much easier

joshdover commented on x-pack/plugins/licensing/server/licensing_plugin.ts:
```diff
@@ -0,0 +1,128 @@
+      new LicensingPluginSetup(null, {}, null, clusterSource)
+    );
+
+    this.poller$ = timer(pollingFrequency);
```
Is this polling logic tested anywhere?

restrry commented:
ack: going to review tomorrow

eliperelman commented:
retest

restrry commented on x-pack/plugins/licensing/server/plugin.ts:
```diff
@@ -0,0 +1,136 @@
+  private create({ clusterSource, pollingFrequency }: LicensingConfig, core: CoreSetup) {
+    this.poller = new Poller<ILicense>(
+      pollingFrequency,
+      new License(null, {}, null, clusterSource),
```
1. signature a bit criptic. does it make sense to pass as object? 
```js
new License({
  license: null,
  features: {},
  error: null,
  clusterSource
});
```
2. What is the point to pass an error?

eliperelman replied:
1. Resolved in next PR.
2. Used by `reasonUnavailable`.

restrry commented on x-pack/plugins/licensing/server/license.ts:
```diff
@@ -0,0 +1,178 @@
+    return LICENSE_TYPE[this.type as LicenseType] >= minimum;
+  }
+
+  check(pluginName: string, minimumLicenseRequired: LICENSE_TYPE | string) {
```
1. why function accepts `string`? do we support licenses with custom name? probably we expect one of the license names and should use `LicenseType`
2. is it possible to narrow down interface to accept either strings or numbers only?

eliperelman replied:
1. This is used for messaging, not anything control-based.
2. The enum is number-based, so yes, strings and numbers.

restrry commented on x-pack/plugins/licensing/server/types.ts:
```diff
@@ -0,0 +1,120 @@
+  /**
+   * Receive a serialized plain object of the license.
+   */
+  toObject(): any;
```
1. it's not DX friendly to return `any`, we know the shape of the object. don't we?
2. Legacy plugin doesn't provide such method.

eliperelman replied:
1. As far as I know, we don't know the complete shape of this.
2. It had a different name prior; this one captures the intent better, improving DX.

restrry commented on x-pack/plugins/licensing/server/license.ts:
```diff
@@ -0,0 +1,178 @@
+export class License implements ILicense {
+  private readonly hasLicense: boolean;
+  private readonly license: any;
+  private readonly features: any;
```
is not `any`, but something with index signature. according to https://github.com/restrry/kibana/blob/ebed1d224a96fd5f91465a5c62a9b266504dc7c7/x-pack/plugins/licensing/server/license.ts#L173

eliperelman replied:
Fixed in next PR.

restrry commented on x-pack/plugins/licensing/server/license.ts:
```diff
@@ -0,0 +1,178 @@
+
+export class License implements ILicense {
+  private readonly hasLicense: boolean;
+  private readonly license: any;
```
we can get type of license from legacy plugin https://github.com/restrry/kibana/blob/ebed1d224a96fd5f91465a5c62a9b266504dc7c7/x-pack/legacy/plugins/xpack_main/server/lib/xpack_info_license.js

eliperelman replied:
Fixed in next PR.

restrry commented on x-pack/plugins/licensing/server/license.ts:
```diff
@@ -0,0 +1,178 @@
+    return this.license.status;
+  }
+
+  public get isActive() {
```
Legacy plugin defined it as a method. should either to follow the same interface or rename the key to avoid confusion?

eliperelman replied:
All of these are properties now.

restrry commented on x-pack/plugins/licensing/server/license_feature.ts:
```diff
@@ -0,0 +1,34 @@
+    return !!this.feature.enabled;
+  }
+
+  public onObject(serializable: LicenseFeatureSerializer) {
```
not used

restrry commented on x-pack/plugins/licensing/server/license_feature.ts:
```diff
@@ -0,0 +1,34 @@
+    isEnabled: this.isEnabled,
+  });
+
+  constructor(public name: string, private feature: any = {}, private license: License) {}
```
- can we define `feature` interface? 
- `license` is not used in the file

eliperelman replied:
1. Fixed in upcoming PR.
2. It is used in serialization.

restrry commented on x-pack/plugins/licensing/server/licensing_config.ts:
```diff
@@ -0,0 +1,23 @@
+  /**
+   * @internal
+   */
+  constructor(rawConfig: LicensingConfigType, env: PluginInitializerContext['env']) {
```
config class is unnecessary if it contains no logic

eliperelman replied:
Config class kept for consistency with other x-pack plugins.

restrry commented on x-pack/plugins/licensing/server/types.ts:
```diff
@@ -0,0 +1,120 @@
+   * A specific API for interacting with the specific features of the license.
+   * @param name the name of the feature to interact with
+   */
+  getFeature(name: string): LicenseFeature | undefined;
```
should always return LicenseFeature. otherwise interface is not friendly. example https://github.com/elastic/kibana/pull/43623/files#diff-ced7be106437c4edb1589d1242111e88R24

eliperelman replied:
Fixed in next PR.

restrry commented on x-pack/plugins/licensing/server/plugin.ts:
```diff
@@ -0,0 +1,136 @@
+      );
+  }
+
+  private hasLicenseInfoChanged(newLicense: any) {
```
can we use something like `lodash.isEqual` or Rx.distinctUntilChanged? and with `any` it's hard to say that all fields are covered by https://github.com/elastic/kibana/pull/43623/files#diff-6cafae74a549dc892111f589f7faa6a7R48-R50

eliperelman replied:
This is custom logic for the previous x-pack changing implementation, so we don't want to deviate in case there are special edge cases.

restrry commented on x-pack/plugins/licensing/server/plugin.ts:
```diff
@@ -0,0 +1,136 @@
+        );
+
+        if (license !== false) {
+          return new License(license, features, error, clusterSource);
```
is it poller responsibility to store value? I'd say plugin should do it

eliperelman replied:
It is the poller's responsibility since it is the owner of the BehaviorSubject.

restrry commented on x-pack/plugins/licensing/server/license.ts:
```diff
@@ -0,0 +1,178 @@
+      return `[${this.clusterSource}] Elasticsearch cluster did not respond with license information.`;
+    }
+
+    if (this.error instanceof Error && (this.error as any).status === 400) {
```
who sets this `status` field? cannot see it on Boom errors

eliperelman replied:
This was a carry-over from the previous implementation, so I'm not sure who sets this. Any recommendations on an alternative?

restrry commented on src/core/utils/poller.ts:
```diff
@@ -0,0 +1,55 @@
+  public readonly subject$ = new BehaviorSubject<T>(this.initialValue);
+  private poller$ = timer(0, this.frequency);
+  private subscription = this.poller$.subscribe(async iteration => {
+    const next = await this.handler(iteration);
```
`subject$` value can have a race condition, because handler is async.

eliperelman replied:
This would be the responsibility of the consumer to avoid race conditions. We cannot say for certain that consumers would want to prevent the next poll because their current handler has not completed.

restrry commented on x-pack/plugins/licensing/server/constants.ts:
```diff
@@ -0,0 +1,21 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export const SERVICE_NAME = 'licensing';
```
not used

eliperelman replied:
Fixed in next PR.

restrry commented on x-pack/plugins/licensing/server/constants.ts:
```diff
@@ -0,0 +1,21 @@
+ */
+
+export const SERVICE_NAME = 'licensing';
+export const DEFAULT_POLLING_FREQUENCY = 30001; // 30 seconds
```
I don't think we should have a dedicated constants file. this value exists only in config context, no one else should read it

eliperelman replied:
This is used internally to the plugin, not external consumers per se.

restrry commented on x-pack/plugins/licensing/server/index.ts:
```diff
@@ -0,0 +1,13 @@
+import { schema } from './schema';
+import { Plugin } from './plugin';
+
+export * from './types';
```
I'd expect `LicenseType`, `LICENSE_TYPE` and other types users interact with to be exported as well

eliperelman replied:
Fixed in next PR.

restrry commented on x-pack/plugins/licensing/server/license.test.ts:
```diff
@@ -0,0 +1,180 @@
+        },
+      }));
+
+      expect(license.check('test', 'basic').check).toBe(LICENSE_STATUS.Valid);
```
interface `check(...).check` looks a bit ambiguous. could we rename it? `check().status` for example

eliperelman replied:
It's called `check` due to `status` already having a meaning on the license.

restrry commented on x-pack/plugins/licensing/server/license.ts:
```diff
@@ -0,0 +1,178 @@
+    return this.objectified;
+  }
+
+  getFeature(name: string) {
```
with immutable `License` class, `LicenseFeature` doesn't track license updates

eliperelman replied:
The `LicenseFeature` class isn't meant to track changes. That is the responsibility of the feature plugin.

eliperelman commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/232/)

spalger commented:
@eliperelman One of the tests from this PR which hasn't been backported yet is flaky, and has been skipped in master, please don't backport the flaky test without the skip, or fix the flaky test before backporting.

eliperelman commented:
@restrry forgot to post here, I will be addressing all of your review comments in my next PR as several of your suggestions have already been written in my follow up. Thanks! :+1:

eliperelman commented on x-pack/plugins/licensing/server/plugin.ts:
```diff
@@ -0,0 +1,130 @@
+  constructor(private readonly context: PluginInitializerContext) {
+    this.logger = this.context.logger.get();
+    this.config$ = this.context.config
+      .create<{ config: LicensingConfigType }>()
```
There is an inconsistency in config creation between testing and production, which I have normalized in the next PR.
